src/sagemaker/estimator.py

```diff
@@ -328,6 +328,17 @@ def prepare_workflow_for_training(self, job_name=None):
         """
         self._prepare_for_training(job_name=job_name)
 
+    def _ensure_base_job_name(self):
+        """Set ``self.base_job_name`` if it is not set already."""
+        # honor supplied base_job_name or generate it
+        if self.base_job_name:
+            return
+
+        if isinstance(self, sagemaker.algorithm.AlgorithmEstimator):
+            self.base_job_name = self.algorithm_arn.split("/")[-1]  # pylint: disable=no-member
```

Comment:
Beyond scope for this PR, but I wonder if you want to model this so it can be resolved as an overridden method in `AlgorithmEstimator` to favor co-locating logic specific to derived classes to those derived classes. I've seen other violations of LSP/sub-type leakage elsewhere in the package, so it may be worth looking into in the future.

Reply:
I'd been annoyed at this one previously, and hadn't changed it before because it was embedded in the method, but now that I'm refactoring it out...